src/gluonnlp/model/transformer.py

```diff
@@ -990,58 +953,180 @@ def decode_seq(self, inputs, states, valid_length=None):
         -------
         output : NDArray, Shape (batch_size, length, C_out)
         states : list
-            The decoder states, includes:
-
+            The decoder states:
             - mem_value : NDArray
-            - mem_masks : NDArray, optional
+            - mem_masks : NDArray or None
         additional_outputs : list of list
             Either be an empty list or contains the attention weights in this step.
             The attention weights will have shape (batch_size, length, mem_length) or
             (batch_size, num_heads, length, mem_length)
         """
-        batch_size = inputs.shape[0]
-        length = inputs.shape[1]
-        length_array = mx.nd.arange(length, ctx=inputs.context, dtype=inputs.dtype)
-        mask = mx.nd.broadcast_lesser_equal(
-            length_array.reshape((1, -1)),
-            length_array.reshape((-1, 1)))
+
+        length_array = F.contrib.arange_like(inputs, axis=1)
+        mask = F.broadcast_lesser_equal(length_array.reshape((1, -1)),
+                                        length_array.reshape((-1, 1)))
         if valid_length is not None:
-            arange = mx.nd.arange(length, ctx=valid_length.context, dtype=valid_length.dtype)
-            batch_mask = mx.nd.broadcast_lesser(
-                arange.reshape((1, -1)),
-                valid_length.reshape((-1, 1)))
-            mask = mx.nd.broadcast_mul(mx.nd.expand_dims(batch_mask, -1),
-                                       mx.nd.expand_dims(mask, 0))
+            batch_mask = F.broadcast_lesser(length_array.reshape((1, -1)),
+                                            valid_length.reshape((-1, 1)))
+            batch_mask = F.expand_dims(batch_mask, -1)
+            mask = F.broadcast_mul(batch_mask, F.expand_dims(mask, 0))
         else:
-            mask = mx.nd.broadcast_axes(mx.nd.expand_dims(mask, axis=0), axis=0, size=batch_size)
-        states = [None] + states
-        output, states, additional_outputs = self.forward(inputs, states, mask)
-        states = states[1:]
+            mask = F.expand_dims(mask, axis=0)
+            mask = F.broadcast_like(mask, inputs, lhs_axes=(0, ), rhs_axes=(0, ))
+
+        mem_value, mem_mask = states
+        if mem_mask is not None:
+            mem_mask = F.expand_dims(mem_mask, axis=1)
+            mem_mask = F.broadcast_like(mem_mask, inputs, lhs_axes=(1, ), rhs_axes=(1, ))
```

Comment:
as in xingjian's slides, we can extract the code of generating mask into a fn.

Reply:
Let's do it in a follow-up PR as it will extend the `attention_cell` API